chainntnfs/txconfnotifier.go

```diff
@@ -0,0 +1,227 @@
+			ntfn.details = &TxConfirmation{
+				BlockHash:   blockHash,
+				BlockHeight: blockHeight,
+				TxIndex:     uint32(tx.Index()),
+			}
+
+			confHeight := blockHeight + ntfn.NumConfirmations - 1
+			tcn.ntfnsByConfirmHeight[confHeight] =
+				append(tcn.ntfnsByConfirmHeight[confHeight], ntfn)
+			tcn.confTxsByInitialHeight[blockHeight] =
+				append(tcn.confTxsByInitialHeight[blockHeight], tx.Hash())
+		}
+	}
+
+	// Dispatch notifications for all transactions that are considered confirmed
+	// at this new block height.
+	for _, ntfn := range tcn.ntfnsByConfirmHeight[tcn.currentHeight] {
+		ntfn.Event.Confirmed <- ntfn.details
+		ntfn.dispatched = true
+	}
+	delete(tcn.ntfnsByConfirmHeight, tcn.currentHeight)
+
+	// Clear entries from confNotifications and confTxsByInitialHeight. We
+	// assume that reorgs deeper than the reorg safety limit do not happen, so
+	// we can clear out entries for the block that is now mature.
+	matureBlockHeight := tcn.currentHeight - tcn.reorgSafetyLimit
+	for _, txHash := range tcn.confTxsByInitialHeight[matureBlockHeight] {
+		delete(tcn.confNotifications, *txHash)
+	}
+	delete(tcn.confTxsByInitialHeight, matureBlockHeight)
+
+	return nil
+}
+
+// DisconnectTip handles the tip of the current chain being disconnected during
+// a chain reorganization. If any watched transactions were included in this
+// block, internal structures are updated to ensure a confirmation notification
+// is not sent unless the transaction is included in the new chain.
+func (tcn *TxConfNotifier) DisconnectTip(blockHeight uint32) error {
+	if blockHeight != tcn.currentHeight {
+		return fmt.Errorf("Received blocks out of order: "+
+			"current height=%d, disconnected height=%d",
+			tcn.currentHeight, blockHeight)
+	}
+	tcn.currentHeight--
+
+	for _, txHash := range tcn.confTxsByInitialHeight[blockHeight] {
+		for _, ntfn := range tcn.confNotifications[*txHash] {
+			confHeight := blockHeight + ntfn.NumConfirmations - 1
+
+			// Find index of ntfn in ntfnsByConfirmHeight slice.
+			idx := -1
+			slice := tcn.ntfnsByConfirmHeight[confHeight]
+			for i, entry := range slice {
+				if entry == ntfn {
+					idx = i
+					break
+				}
+			}
+
+			// Remove ntfn from ntfnsByConfirmHeight slice.
+			if idx == -1 {
+				continue
+			} else if len(slice) == 1 {
+				delete(tcn.ntfnsByConfirmHeight, confHeight)
+			} else {
+				// See https://github.com/golang/go/wiki/SliceTricks#delete-without-preserving-order
+				slice[idx] = slice[len(slice)-1]
+				slice[len(slice)-1] = nil
+				tcn.ntfnsByConfirmHeight[confHeight] = slice[:len(slice)-1]
+			}
+		}
+	}
+	delete(tcn.confTxsByInitialHeight, blockHeight)
+
+	return nil
+}
+
+// TearDown is to be called when the owner of the TxConfNotifier is exiting.
+// This closes the event channels of all registered notifications that have
+// not been dispatched yet.
+func (tcn *TxConfNotifier) TearDown() {
+	for _, ntfns := range tcn.confNotifications {
```

Comment:
This should also close a quit channel localized to the `TxConfNotifier`. All sends above would select on this quit channel to ensure things can be shutdown gracefully.